proposals/NNNN-specialize.md

```diff
@@ -0,0 +1,193 @@
+# Explicit Specialization
+
+* Proposal: [SE-NNNN](NNNN-specialize.md)
+* Authors: [Ben Cohen](https://github.com/airspeedswift)
+* Review Manager: TBD
+* Implementation: Available in nightly toolchains using the underscored `@_specialize`
+* Status: **Awaiting review**
+
+## Introduction
+
+The Swift compiler has the ability to "specialize" a generic function at compile time. This specialization creates a custom implementation of the function, where the generic placeholders are substituted with specific types. This can unlock optimizations of that specialized function that can be dramatically faster than the unspecialized version in some circumstances. The compiler can generate this specialized version and call it when it can see at the call site with what concrete types a function is being called, and the body of the function to specialize.
+
+In some cases, though, this information is obscured from the compiler. This proposal introduces a new attribute, `@specialize`, which allows the author of a generic function to generate pre-specialized versions of that function for specific types. When the unspecialized version of the function is called with one of those types, the compiler will generate code that will re-dispatch to those prespecialized versions if available.
+
+## Motivation
+
+Consider the following generic function that sums an array of any binary integer type:
+
+```swift
+extension Sequence where Element: BinaryInteger {
+  func sum() -> Double {
+    reduce(0) { $0 + Double($1) }
+  }
+}
+```
+
+If you call this function directly on an array of a specific integer type (e.g. `Array<Int>`):
+
+```
+let arrayOfInt: [Int] = // ...
+let result = arrayOfInt.sum()
+```
+
+in optimized builds, the Swift compiler it will generate a specialized version of `sum` for that type. 
+
+If you inspect the binary, you will see this specialized version under a symbol `_$sST3sumSz7ElementRpzrlEAASdyFSaySiG_Tg5`, which demangles to `generic specialization <[Swift.Int]> of (extension in example):Swift.Sequence< where A.Element: Swift.BinaryInteger>.sum() -> Swift.Double`, alongside the unspecialized version.
+
+This specialized version of `sum` will be optimized specifically for `Array<Int>`. It can move a pointer directly over the the array's buffer, loading the elements into registers directly from memory. On modern hardware, it can make use of dedicated instructions to convert the them to floating point. It can do this because it can see the implementation of `sum`, and can see the exact types on which it is being called. What exact assembly instructions are generated will differ significantly between e.g. `Array<Int8>.sum` and `Array<UInt64>.sum`.
```

Comment:
...to convert them...

Reply:
```suggestion
This specialized version of `sum` will be optimized specifically for `Array<Int>`. It can move a pointer directly over the the array's buffer, loading the elements into registers directly from memory. On modern hardware, it can make use of dedicated instructions to convert the integers to floating point. It can do this because it can see the implementation of `sum`, and can see the exact types on which it is being called. What exact assembly instructions are generated will differ significantly between e.g. `Array<Int8>.sum` and `Array<UInt64>.sum`.
```